Ccdb fetcher

ktf opened the pull request:
Ongoing:

- [x] Understand why my changes to the CCDBApi seem to crash some other tests.
- [x] Add the fetching of the CTP and associated calculation of the timestamp based on OrbitResetTime and firstTFOrbit
- [x] Do not deserialise the objects every timeframe
- [x] Add a callback when it actually needs to happen.
- [x] Use DISTSUBTIMEFRAME / TF/TFNUM where appropriate to get the time for the data
- [x] Add ConfigParamSpec for createdNotAfter and createdNotBefore
- [x] Support for metadata on query in the InputSpec
- [x] Allow configuring the CCDB path / backend on a per input basis.
- [ ] Support for headers as result message.
- [ ] Better strategy to cleanup the cache and eventually support prefetching of CCDB objects / multiple in-fly objects for the same path.

ktf committed:
DPL: attempt at proper implementation for CCDBFetcher

ktf commented:
First iteration, do not merge.

shahor02 commented on Framework/Core/src/CCDBHelpers.cxx:
```diff
@@ -12,24 +12,110 @@
+
+            LOGP(info, "{} {} {}\n", timingInfo.timeslice, timingInfo.tfCounter, timingInfo.creation);
+            helper->api.loadFileToMemory(v, path, metadata, timingInfo.timeslice, &headers, etag, createdNotAfter, createdNotBefore);
+            if ((headers.count("Error") != 0) || !v.size()) {
```
I think `v.size()==0` should be treated as an error only if the ETAG was empty (cache was not filled yet), with ETAG for the valid object the `v` must be empty.

ktf replied:
agreed.

ktf replied:
done.

ktf commented:
With the latest commit commit the example in `Framework/TestWorkflows/src/test_CCDBFetcher.cxx` works as before when running with:

```
stage/bin/o2-testworkflows-test-ccdb-fetcher --start-value-enumeration `date +%s` --condition-backend http://alice-ccdb.cern.ch:8080
```

but using the new centralised logic and honouring the etag.

shahor02 commented:
>> Understand why my changes to the CCDBApi seem to crash some other tests.

I think this is because of exposing boost headers to root cling. Perhaps we can avoid this by templating the vector to fill?

ktf commented:
I was thinking about that, but then we would need to expose everyone a  bunch of ROOT headers. I am now trying something  slightly different with `ifndef __CLING__`.

ktf committed:
DPL: port OrbitResetTime code

Now supports fetching the CTP centrally, so that it can be used
both for correct calculation of the timestamp from the firstTForbit
and it can be provided as an Input to whoever neeeds it.

ktf commented:
New version fetches the OrbitResetTime from the CTP (plus a bit of cleanup).

ktf committed:
DPL: deserialised conditions are now cached

ktf commented:
Update now does not deserialise CCDB objects if the associated blob was not changed.

ktf committed:
DPL: invoke callback on deserialization

ktf commented:
Now with the callback for deserialised objects.

shahor02 commented on Framework/Core/include/Framework/Task.h:
```diff
@@ -38,6 +38,24 @@ class has_endOfStream
 
+/// Check if the class task has EndOfStream
+template <typename T>
+class has_finalseCCDB
```
typo in "finaliseCCDB"?

ktf replied:
yes

ktf committed:
DPL: adapt to new InputRecord API

ktf committed:
DPL: select correct input for ccdbFetcher

Depending on their availability, this will subscribe TFN/TFNumber
or FLP/DISTSUBTIMEFRAME to the ccdbFetcher so that the timing information
will be extracted from there.

ktf commented:
Fixed some tests and update to include DISTSUBTIMEFRAME / TF/TFNUM subscription, if present in the topology.

ktf committed:
DPL: CCDB support for not created after / before

ktf committed:
DPL: support specifying metadata for CCDB

ktf committed:
DPL: late add of CTP output

ktf commented:
Metadata support added.

ktf committed:
DPL: allow remapping of urls via --condition-remap

Syntax is:

--condition-remap <prefix1>=<target1>,[<target2>;<prefix2>=<target3>[,...]]

ktf commented:
I now have the remapping in place. There is probably something still to adapt for the multihost case, but it should work for simple, same server remappings. The syntax is:

```
--condition-remap <prefix1>=<target1>,[<target2>;<prefix2>=<target3>[,...]]
```

where prefix can be either `https://` / `http://` or `/`. The targets are the paths which are specified by the `ccdbParamSpec()`. For the time being the must fully match.

If this works I would merge this PR and address the remaining issues in separate developments.

shahor02 commented:
> For the time being they must fully match.

If you refer to URLs with the wildcard, I would not do that as this could be error prone.

ktf commented:
@davidrohr @shahor02 I think this could be merged as is. It is a step forward the current implementation and if the Lifetime::Condition is not used, there should not be any side effects at all. What do you think? The FLP suite will use in any case the tag of yesterday, no?

shahor02 commented:
Hi @ktf 
thanks, fine with me to merge this. I'll try to play with it after finishing another branch I am working with.

shahor02 commented:
@ktf were not you going to merge this PR? Is there any reason to not do this?

davidrohr commented:
I discussed with @ktf to check this in FST with DD before merging, this test is currently running. Will comment here soon.

davidrohr reviewed (approved):
Don't see any problem in FST, fine for me to merge this.

ktf commented:
Ok, thanks. As discussed, let's take it from here for the remaining bugs / features.
